internal/cli/capact/build.go

```diff
@@ -0,0 +1,193 @@
+		Dir:    ".",
+		Target: "terraform-runner",
+	},
+	{
+		Name:   "populator",
+		Dir:    ".",
+		Target: "generic-alpine",
+	},
+	{
+		Name:   "e2e",
+		Dir:    ".",
+		Target: "e2e",
+		ExtraBuildArgs: []string{
+			"BUILD_CMD=go test -v -c",
+			"SOURCE_PATH=./test/e2e/*_test.go",
+		},
+	},
+}
+
+func buildImage(w io.Writer, img image, repository, version string) (string, error) {
+	// docker build --build-arg COMPONENT=$(APP) --target generic -t $(DOCKER_REPOSITORY)/$(APP):$(DOCKER_TAG)
+	imageTag := fmt.Sprintf("%s/%s:%s", repository, img.Name, version)
+
+	// #nosec G204
+	cmd := exec.Command("docker",
+		"build",
+		"--build-arg", fmt.Sprintf("COMPONENT=%s", img.Name),
+		"-t", imageTag, ".")
+
+	if img.Target != "" {
+		cmd.Args = append(cmd.Args, "--target", img.Target)
+	}
+
+	cmd.Env = os.Environ()
+	cmd.Env = append(cmd.Env, []string{
+		// enable module support across all go commands.
+		"GO111MODULE=on",
+		// enable consistent Go 1.12/1.13 GOPROXY behavior.
+		"GOPROXY=https://proxy.golang.org",
+	}...)
+
+	if !img.DisableBuildKit {
+		cmd.Env = append(cmd.Env,
+			// enable the BuildKit builder in the Docker CLI.
+			"DOCKER_BUILDKIT=1")
+	}
+
+	if len(img.ExtraBuildArgs) != 0 {
+		for _, arg := range img.ExtraBuildArgs {
+			cmd.Args = append(cmd.Args, "--build-arg", arg)
+		}
+	}
+	cmd.Dir = img.Dir
+	cmd.Stdout = w
+	cmd.Stderr = w
+	err := cmd.Run()
+	if err != nil {
+		return "", err
+	}
+	return imageTag, nil
+}
+
+// DeleteImage deletes passed image
+func DeleteImage(images []string) error {
+	// #nosec G204
+	cmd := exec.Command("docker",
+		"rmi", images[0]) //TODO
+	stdoutStderr, err := cmd.CombinedOutput()
+	fmt.Printf("%s\n", stdoutStderr)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// BuildImages builds passed images setting passed repository and version
+func BuildImages(w io.Writer, repository, version string, names []string) ([]string, error) {
+	imagesMap := mappedImages()
+	var created []string
+
+	for _, name := range names {
+		image, ok := imagesMap[name]
+		if !ok {
+			return nil, fmt.Errorf("can not find image %s", name)
+		}
+		imageTag, err := buildImage(w, image, repository, version)
+		if err != nil {
+			return nil, errors.Wrapf(err, "while building image %s", image.Name)
+		}
+		created = append(created, imageTag)
+	}
+	return created, nil
+}
+
+// SelectImages returns a list of images calculated from focus and skip lists
+func SelectImages(focus, skip []string) ([]string, error) {
+	if len(focus) > 0 && len(skip) > 0 {
+		return nil, errors.New("can not skip and focus images at the same time")
+	}
+
+	imagesMap := mappedImages()
+
+	var selected []string
+	if len(focus) > 0 {
+		for _, name := range focus {
+			_, ok := imagesMap[name]
+			if !ok {
+				return nil, fmt.Errorf("focused image does not exist: %s", name)
+			}
+			selected = append(selected, name)
+		}
+		return selected, nil
+	}
+
+	for _, image := range images {
+		if shouldSkipImage(image.Name, skip) {
```

Comment:
It is only up to 10 elements. I would argue that this approach is noticeable faster :)
Using simple  slices for me is more readable. I like the idea about `slices.go` Maybe I will create a follow-up issue with `low-hanging-fruit` so it can be picked by some new folks?

Reply:
Yep, it's not for making that faster, but to make it consistent and reusable. Currently, we have duplicated funcs for the same purpose. From the new `IndexMap` type and the methods it has, you know already the purpose of it. It can be reused across all places with kinda readable API `opts.SkipComponents.Contains(component.Name())`. We can do it later as `low-hanging-fruit` but definitely we should address it to ensure easy maintainability.